Add Support for Custom Rotation Mappings

KAMKEEL opened the pull request:
Rotation Mappings is a system that generates mapping JSONs in the WorldEdit Config when the world is loaded with a modded configuration of blocks. Vanilla WorldEdit only knows how to rotate vanilla metadata, so on a modded server every mod stair, log, trapdoor and gate comes out facing the wrong way after a `//rotate`. This system fixes that.

On world load the block registry is scanned and every non-vanilla block is sorted by its block class into one of the mapping files (`stairs.json`, `pillar.json`, `trap_door.json`, `fence_gate.json`, `other.json`) inside the WorldEdit config folder. Defaults are generated automatically, and anything the auto-detection misses can be added by hand. `//reloadmappings` (perm `worldedit.reload`) reloads the files live, no restart needed.

Supports:
- Trapdoors
- Buttons (Sometimes they pop off)
- Fence Gates
- Pillars (Logs, Barrels, etc)
- Stairs
- Other (Allows you to apply your own mappings based on those types)

Here is a video: 
[Test with Auto-Generate Default Mod Mappings on a Modded CrucibleMC Server with 100+ Mods]

https://github.com/user-attachments/assets/beb1c657-5474-4e69-88f0-37e7660dab89


Some examples. These JSONs can be manually modified to support other groups of metadata not auto-generated by the Mod Rotation Defaults.

Pillars use named groups of metas, one group per variant. Each group is ordered `[y, x, z]`, so `[0, 4, 8]` means meta 0 is the vertical log, 4 is east-west and 8 is north-south:

```json
// Rotation mappings for pillar
{
  "uptodate:bone_block": {
    "groups": {
      "0": [0, 4, 8],
      "1": [1, 5, 9]
    }
  },
  "uptodate:stripped_log_oak": {
    "groups": {
      "0": [0, 4, 8],
      "1": [1, 5, 9]
    }
  }
}
```

Trapdoors are split into top/bottom halves with their open and closed meta sets, four facings each:

```json
"uptodate:trap_door_iron": {
  "top": {
    "closed": [8, 9, 10, 11],
    "open": [12, 13, 14, 15]
  },
  "bottom": {
    "closed": [0, 1, 2, 3],
    "open": [4, 5, 6, 7]
  }
}
```
  
Stairs use `top` and `bottom` facing arrays, fence gates and buttons use a single four-direction cycle. Extra bits outside the mapped metas (species, pressed, open and so on) are masked off and carried through untouched, which is why one pillar entry can cover a whole log family.

These mappings allow modded block rotations to be supported and added into WorldEdit.

### Flip support

`//flip` is covered as well, not just `//rotate`. The old approach of pulling a yaw angle out of the transform falls apart on mirrors (a flip is not a rotation), so instead each type takes the direction the block is facing, pushes it through the actual transform, and picks the closest resulting direction. That works the same whether the transform is a rotation, a flip, or both stacked together.

- Stairs and trapdoors mirror their facing on horizontal flips. An up/down flip swaps regular and upside-down variants.
- Fence gates, buttons and other four-direction blocks mirror their facing. Open/pressed bits are kept.
- Pillars keep their axis on flips (an upside-down log is still a vertical log) and still swap x/z under rotation.

### Also in this branch

- `/fixlighting` commands with an async lighting repair queue. Only one repair job can be queued at a time so spamming the command cannot pile up work. Needed some small `World`/`AbstractWorld` API additions.
- Fixed `AddBlocks2` reconstruction when loading schematics. The high nibbles were shifted into bits 8-11 instead of 12-15, so NotEnoughIDs block IDs above 4095 loaded as a completely different block. The combine logic now lives in its own `combineBlockIds` method with a unit test.
- Schematic loading honours `WEOffset` even when `WEOriginX` is missing, so schematics converted from modern formats keep their paste anchor.

### Testing
`RotationMappingsTest` has 32 cases running every rotation type through all four facings, and the new `RotationFlipTest` adds 12 more: both horizontal flip axes, vertical flips, state bit preservation, and flip combined with rotation (including checking that flipping both horizontal axes equals a 180 degree rotation). On top of the unit tests, this has been running on a live CrucibleMC 1.7.10 server with 100+ mods (see the video above), and NEID schematics paste with the correct block IDs.

### Known limitations

- Doors are left alone on purpose. They are two-block states and deserve their own follow-up rather than a half solution.
- Buttons can occasionally pop off after a transform if they end up without a supporting block behind them, that is the block update doing its thing rather than a mapping problem.
- Forge Microblocks / Carpenter's blocks keep their existing compat path. Custom mapping support for those could come later.

KAMKEEL committed:
feat: rotate modded pillar and stair blocks

KAMKEEL committed:
Handle modded doors and trapdoors

KAMKEEL committed:
Fix rotation direction and stair variant handling

KAMKEEL committed:
Fix door rotation orientation

KAMKEEL committed:
Change Depen

KAMKEEL committed:
Add configurable rotation mappings and tests

KAMKEEL committed:
Merge pull request #2 from KAMKEEL/codex/add-custom-rotation-mappings-for-modded-blocks

Add configurable mod rotation mappings

KAMKEEL committed:
Initialize mod rotation mappings on server start

KAMKEEL committed:
Refactor mod rotation mappings

KAMKEEL committed:
Add comments to rotation mapping files

KAMKEEL committed:
Fix door rotation fallback

KAMKEEL committed:
Expand default mappings and tests

KAMKEEL committed:
Fix door and pillar rotation logic

KAMKEEL committed:
Fix door meta defaults

KAMKEEL committed:
Improve pillar transform detection

KAMKEEL committed:
Fix stair mapping and add button rotation

KAMKEEL committed:
Expand rotation unit tests

KAMKEEL committed:
Revamp rotation mappings and tests

KAMKEEL committed:
Fix stair mapping and document custom mapping types

KAMKEEL committed:
Add door and trapdoor rotation mappings

KAMKEEL committed:
Fix orientation detection for rotations

KAMKEEL committed:
Merge branch '3bgmrh-codex/add-custom-rotation-mappings-for-modded-blocks'

KAMKEEL committed:
Expand rotation tests

KAMKEEL committed:
Initialize mod rotation config and handle null

KAMKEEL committed:
Merge mod rotation config into mappings

KAMKEEL committed:
Remove door rotation support and enhance trapdoor mappings

KAMKEEL committed:
Fix trapdoor rotation steps

KAMKEEL committed:
Refactor

KAMKEEL committed:
GameData for Servers

KAMKEEL committed:
Apply Spotless

KAMKEEL committed:
Fix Star Imports

KAMKEEL committed:
More Star Import Fixes

KAMKEEL commented:
Just added video to help demonstrate it! Sorry the quality is very compressed!

Inphysible commented:
Woah, this would be super convenient. Does //flip work for these items as well? And, as for "special" modded items, would that include anything from Forge Microblocks?

KAMKEEL commented:
@Inphysible No, I haven't tweaked Flip yet, unfortunately. Only dealing with rotations first. I'll probably add Flip Support on later date. Flip Upside shouldn't work. But Flip Horizontally - Uses 2 different rotation handlers, so in theory it should also work.

KAMKEEL commented:
@Inphysible No microblocks yet, but I'm sure I can build it into the Microblocks handler

KAMKEEL committed:
Add additional pillar defaults and update JSON headers

KAMKEEL committed:
Add command to reload rotation mappings

KAMKEEL committed:
Merge pull request #6 from KAMKEEL/codex/add-meta-2-and-3-as-defaults

Expand pillar defaults and update mapping comments

KAMKEEL committed:
Merge pull request #7 from KAMKEEL/codex/add-command-to-reload-rotation-mappings

Add rotation mapping reload command

KAMKEEL committed:
Spotless

KAMKEEL committed:
fix: optimize lighting repair and rotation handling

KAMKEEL committed:
feat: add lighting manipulation commands

KAMKEEL committed:
Enforce single queued lighting repair

KAMKEEL committed:
Merge pull request #10 from KAMKEEL/codex/improve-async-execution-for-fix-lighting-command

Disallow concurrent fix lighting scheduling

KAMKEEL committed:
Fix Spotless

boubou19 commented:
@KAMKEEL any news? it does not build

Dream-Master committed:
Merge branch 'master' into rotations

Dream-Master committed:
Merge branch 'master' into rotations

KAMKEEL committed:
Merge remote-tracking branch 'origin/master' into neid-build

KAMKEEL committed:
Fix AddBlocks2 block ID reconstruction

KAMKEEL committed:
Honour WEOffset when WEOriginX is absent

Converted schematics carry the paste anchor as WEOffset without a
WEOrigin. Reading the two tag groups independently keeps the paste
anchor working for those files instead of silently ignoring it.

KAMKEEL committed:
Remove development notes from docs

KAMKEEL committed:
Support //flip in the rotation mapping system

Flips are affine transforms with a mirrored axis. The yaw based
transform fallback misread mirrors as rotations, so flipped mod blocks
came out with arbitrary facings.

- Stairs and trapdoors: horizontal flips resolve through direction
  projection, vertical flips swap the regular and upside down halves
- Pillars: transform now projects the pillar axis through the actual
  transform (nearest axis by absolute dot product) instead of deriving
  a yaw angle, handling rotations and flips alike
- Trapdoors: transform projects the attachment wall direction instead
  of deriving a yaw angle
- Fence gates and buttons already projected directions and now have
  flip coverage locked in by tests
- New RotationFlipTest covers every type across both horizontal axes,
  vertical flips, state bit preservation (open/pressed/species) and
  flip and rotation composition

spacebuilder2020 reviewed (commented):
Code looks logically sound, and tests are passing.  I am holding back on approving until either I can test it myself or another user can test it and make sure that all of the new functions work correctly.  (i.e. now issues happen when fixing lighting, rotations work as expected on supported block types, etc)

spacebuilder2020 commented on src/main/java/com/sk89q/worldedit/extent/clipboard/io/SchematicReader.java:
```diff
@@ -289,64 +295,77 @@ public CompoundTag apply(CompoundTag nbtData) {
-                                    if (nbtData.containsKey("tag") && itemMap.get("tag") instanceof CompoundTag nbt) {
-                                        itemMap.put("tag", apply(nbt));
+                                    if (nbtData.containsKey("tag") && itemMap.get("tag") instanceof CompoundTag) {
+                                        itemMap.put("tag", apply((CompoundTag) itemMap.get("tag")));
```
Is there a reason you replaced the instanceof cast with a second cast below?  The main reason the cast was above was to reduce duplication and help slightly with performance.

spacebuilder2020 replied:
Marking this as resolved.  I took a look and determined that this feature is a new java feature and doesn't compile on Java 8.

spacebuilder2020 commented on src/main/java/com/sk89q/worldedit/extent/clipboard/io/SchematicReader.java:
```diff
@@ -289,64 +295,77 @@ public CompoundTag apply(CompoundTag nbtData) {
-                                    if (nbtData.containsKey("d") && itemMap.get("d") instanceof CompoundTag d) {
-                                        itemMap.put("d", apply(d));
+                                    if (nbtData.containsKey("d") && itemMap.get("d") instanceof CompoundTag) {
+                                        itemMap.put("d", apply((CompoundTag) itemMap.get("d")));
```
Same as above

spacebuilder2020 replied:
Marking this as resolved.  I took a look and determined that this feature is a new java feature and doesn't compile on Java 8.

KAMKEEL replied:
OH I JUST REMOVED IT

KAMKEEL replied:
Okay, I removed the removed it

spacebuilder2020 reviewed (changes requested):
I want to make a note of this since this was from merging in changes from another feature and not part of the primary feature for this PR.

Note: it was my code, so I don't mind you changing some of it and it looks like you refactored a bit of it and it is cleaner, but I am not sure on this part here.

KAMKEEL committed:
Update SchematicReader.java

KAMKEEL committed:
Revert "Update SchematicReader.java"

This reverts commit da5a70ea780787d472e66ca781b56584c9c5d6bd.

spacebuilder2020 reviewed (approved):
After a second review, and some quick research and testing, I have determined that my prev request changes was incorrect due to the change being about a new Java feature that was inadvertently slipped into the code and the change was correcting that issue.
